### PR TITLE
[Don't Merge] parallax code autopsy

### PR DESCRIPTION
**Pull request by DeityLink**
It's unfinished, it's laggy, I'm only making this PR so other coders can help me make it run fine. I'll make another PR later down the line to actually get it merged.

http://i.4cdn.org/vg/1463351337179.webm

I gave a go at parallax over a year ago but gave up.

Then BYOND 510 came about and someone over at Paradise actually did show that it was possible ParadiseSS13/Paradise#4085, then @dylanstrategie tried porting it #9884 but I didn't like his port (which he gave up due to feedback) and I'm lukewarm on Paradise's single-layered parallax. So I spent hours making 4 versions of every icon_state in turf/space.dmi, to bring you this versions.

Pros:
- Multi-layered (static background+layer that moves at 1 pixel per movement+layer that moves at 2 pixels per movement+dust that moves along with the turfs)
- Players can tweak it in their preferences (enable it, disable it completely, have all the layers be static) They can as well turn off the dust.
- Since the stars are overlay images instead of being blended/static images, they keep their twinkling animations.
- Position of the parallax layers is tied to player movement rather than position in the Z level, which means that drifting from a Z Level to the next won't cause a "Jump" in the parallax. The layers loop around when the player goes too far in the same direction.

Cons:
- Currently causes abysmal lag so I'm currently optimizing it all.

### CODE REVIEW

**Comment by Brolaire** [comment]
What does this do?

**Comment by Imagebard** [comment]
Seems to be something that makes the space tiles look like the same for other people. Also some "Star Dust" thing I don't really understand.

**Comment by PJB3005** [comment]
@Brolaire it makes the thread furious.

**Comment by PJB3005** [comment]
http://i.4cdn.org/vg/1463351596433.webm

**Comment by Imagebard** [comment]
First the lightning changes and now this

party

**Comment by DeityLink** [comment]
pinging @monster860 whose code was kinda used as a base for this. you might want to check this out.

**Comment by monster860** [comment]
The performance of `get_mobs_in_contents()` proc has not been tested and could cause performance issues with lots of players. I would recommend testing this yourself or waiting until the Paradise maintainers do.

**Comment by monster860** [comment]
Another thing I would like ro mention, but it feels as if the stars are somewhat lacking in color in the foreground.

**Comment by DeityLink** [comment]
@monster860 Space in SS13 isn't completely black, it's using color #151515 while here I tried making a darker-looking space by having the canvas (the parallax_void) use color #000000.

Also recording with Fraps+Converting into webm causes some color loss. Check the webm in the top post rather than the one posted by PJB, since that one is at higher bitrate.

File: code/game/atoms_movable.dm [hunk 1]
```diff
@@ -166,6 +166,9 @@
 					else if (step(src, WEST))
 						. = step(src, SOUTH)
 
+	for(var/mob/M in get_mobs_in_contents())
```
Comment (d3athrow) [inline code/game/atoms_movable.dm:169]: hooks seem like a better thing for this.

**Comment by DeityLink** [comment]
Ok so the good news is that I found the source of the lag and it had nothing to do with neither updating the layers, or get_mobs_in_contents(). So that's one problem solved.

The bad news is that because I wasn't using blend_mode = BLEND_MULTIPLY the layers appeared over hidden turfs: 
shadows

And currently the only workaround I've found is to have space be white:
whitespace

**Comment by monster860** [comment]
You need to put the canvas on the same `plane` as the space turfs, give them normal texture, set `appearance_flags` of the camvas to PLANE_MASTER, and set `color` to `list(0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,0)`.

**Comment by monster860** [comment]
Ignore that above comment Im full of shit there.

Put all the space layers on the same plane, use the overlay blend for them, and have a BLEND_MULTIPLY PLANE_MASTER layer on the same plane

**Comment by DeityLink** [comment]
@monster860  What does PLANE_MASTER do exactly?

So I put both all the /obj/screen/parallax, as well as the /obj/screen/parallax_void and the /obj/screen/parallax_canvas on plane -9 (space turfs are plane -10) (all have layer = 1)

the /obj/screen/parallax and the /obj/screen/parallax_canvas have blend_mode = BLEND_OVERLAY, while the /obj/screen/parallax_void has blend_mode = BLEND_MULTIPLY and appearance_flags = PLANE_MASTER.

the /obj/screen/parallax_canvas's color was set to black, and this was the result

(the space dust is a turf overlay that's on a higher plane)

So then I thought that the parallax_canvas might be appearing above the stars, so I tried either lowering its layer to 0 or its plane to -10 but I'd end up with it looking like this again:

**Comment by monster860** [comment]
Keep parallax void as it was before. Make another layer with plane -9 and BLEND_MULTIPLY.

**Comment by monster860** [comment]
Another comment, I oroginally opened a pr in February for 509 parallax. So it was possible then.

**Comment by DeityLink** [comment]
@monster860 I'm intrigued by the for(var/client/C in parallax_on_clients) C.images |= A.white_overlay

wouldn't that make a lot of images? And how would you update that?

**Comment by monster860** [comment]
That was my way of making space white for parallax clients only.

I don't do this anymore, I have a PLANE_MASTER plane.

**Comment by Intigracy** [comment]
why'd you get rid of the red dots in space wtffffffff

**Comment by DeityLink** [comment]
@monster860 Alright then please tell me, how do you make your Plane_Master plane only appear over space and not, let's say, behind walls, where players aren't supposed to see?

**Comment by DeityLink** [comment]
@Intigracy but I didn't

**Comment by monster860** [comment]
@DeityLink

Here's what you need
- Space turfs should have plane -10
- There should be a `appearance_flags = PLANE_MASTER` layer with color set to `list(0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1)` (Something like that I think) on plane -10, and blend_mode set to BLEND_OVERLAY
- A `appearance_flags = PLANE_MASTER` layer with blend_mode set to BLEND_MULTIPLY on plane -9
- Your various parallax layers that contain your actual star textures, where the background is BLACK, and `blend_mode = BLEND_ADD`, and `plane = -9`

That right there should be what you need.

**Comment by DeityLink** [comment]
ohhh, so I need a black background instead of a transparent one? geez, this is gonna take some more time in Gimp.

**Comment by ghost** [comment]
I've went through the exact same problem

You need to blend_multiply with a white background so it doesn't try to blend over the black void, funnily enough exadv1 out of all people ran into this problem when trying to add an actual parallax forever ago

**Comment by DeityLink** [comment]
@dylanstrategie You don't need a white background at all, nor a black void. As long as you used BLEND_ADD there won't be issues.

**Comment by monster860** [comment]
@DeityLink You need the white background, otherwise it will show up in the obscured areas.

**Comment by DeityLink** [comment]
ALRIGHT FOUND THE PROBLEM

For some reason, every time you change mob, your master planes' appearance_flags return to 0. I just have to add a check for that.

**Comment by clusterfack** [comment]
If its a pooled hud element thats because there is a bug floating around 510 that was only just recently fixed that I reported here http://www.byond.com/forum/?post=2080853.

**Comment by DeityLink** [comment]
Alright, I've managed to get everything working as intended. I'm going to close this PR now and open another one.

**Comment by PJB3005** [comment]
😑

**Comment by clusterfack** [comment]
Why though

**Pull request closed without merging**